src/libertem/executor/delayed.py

```diff
@@ -0,0 +1,417 @@
+            udf.set_views_for_partition(task.partition)
+            dest = udf.results.get_proxy()
+            merged = MergeAttrMapping(renested)
+            for k in dest:
+                getattr(dest, k)[:] = getattr(merged, k)
+
+        v = damage.get_view_for_partition(task.partition)
+        v[:] = True
+        return (udfs, damage)
+
+    def _accumulate_part_results(self, udf, part_results, task):
+        """
+        If the udf has a merge_all method, this function is used
+        to accumulate dask array-backed partial results on the executor
+        so that the merge_all can be called on all of them at once
+
+        Ensures that the results are correctly ordered and complete
+        before allowing merge_all to be called
+        """
+        buf = next(iter(part_results.values()))  # get the first buffer
+        slice_with_roi = buf._slice_for_partition(task.partition)
+        self._part_results[udf][slice_with_roi] = part_results
+
+        # number of frames in dataset
+        if udf.meta.roi is not None:
+            target_coverage = np.count_nonzero(udf.meta.roi)
+        else:
+            target_coverage = prod(task.partition.meta.shape.nav)
+        # number of frames we have results for
+        current_coverage = sum(prod(k.shape.nav) for k in self._part_results[udf].keys())
+        if target_coverage == current_coverage:
+            ordered_results = sorted(self._part_results[udf].items(),
+                                     key=lambda kv: kv[0].origin[0])
+            self._part_results[udf] = OrderedDict(ordered_results)
+            return True
+        elif current_coverage > target_coverage:
+            raise RuntimeError('More frames accumulated than ROI specifies - '
+                              f'target {target_coverage} - processed {current_coverage}')
+        return False
+
+    def results_for_dataset_sync(self, dataset: DataSet, executor: 'DelayedJobExecutor',
+            roi: Optional[np.ndarray] = None, progress: bool = False,
+            corrections: Optional[CorrectionSet] = None, backends: Optional[BackendSpec] = None,
+            dry: bool = False) -> Iterable[tuple]:
+
+        executor.register_master_udfs(self._udfs)
+
+        return super().results_for_dataset_sync(
+            dataset, executor, roi=roi, progress=progress,
+            corrections=corrections, backends=backends, dry=dry
+        )
+
+    def run_for_dataset_sync(self, dataset: DataSet, executor: JobExecutor,
+            roi: Optional[np.ndarray] = None,
+            progress: bool = False, corrections: Optional[CorrectionSet] = None,
+            backends: Optional[BackendSpec] = None, dry: bool = False,
+            iterate: bool = True) -> Generator["UDFResults", None, None]:
+        if iterate:
+            raise NotImplementedError(
+                "Iterating over partial results is currently not supported "
+                "with the DelayedJobExecutor."
+            )
```

Comment:
For this and the following comment for the async case I don't really understand the use case. It just doesn't make sense to me to iterate over delayed partial results or generate them asynchronously. Do you have any proposal? For that reason I'd rather leave it unimplemented until there's a clear application instead of adding sth to the public API that we'll have to support in the future and that may or may not make any sense. ALso, that saves us all the testing code for it as well. I'd prefer to add things only with a clear use case, not just for completeness sake

Reply:
Why does this need changes to the public API? The `Context` already supports running in sync/async iter/non-iter combinations, for example:

```python
from libertem.api import Context
from libertem.executor.delayed import DelayedJobExecutor
from libertem.udf.sum import SumUDF

ctx = Context(executor=DelayedJobExecutor())
ds = ctx.load("auto", path="...")
res = await ctx.run_udf(dataset=ds, udf=SumUDF(), sync=False)  # errs our with the above `NotImplementedError`
for res in ctx.run_udf_iter(dataset=ds, udf=SumUDF()):
    pass  # similarly errs out
```

Currently, this means that `DelayedJobExecutor` cannot be used in places where other executors just work. Specifically, think about integration of LiberTEM into other software - if it is written in an `async`/`await`-style, it will probably also use the `async` API of LiberTEM. I would prefer if things just work, vs having to maintain a support matrix for executors.

Reply:
My memory is a bit fuzzy, but I recall that I realized "hey, we should at least test the full API", and it didn't work out of the box. We clearly communicate that the DelayedJobExecutor is experimental, for that reason I just raise `NotImplementedError` for the cases that are not well-supported yet.

Yes, I agree for all other executors that we should support everything. The `DelayedJobExecutor` is just an odd one in the mix for the special task of generating Dask arrays. It would break an embedding application anyway, resp. require special handling of the result.

Reply:
TL; DR I'd strongly prefer the `NotImplementedError` over additional work to support iteration and async at this time.

Reply:
This is done in 4e8d168b - according to the test cases, it was enough to remove the exceptions thrown in `run_for_dataset_{a,}sync` (and it is only supported on Python 3.7).